app/services/bike_finder.rb

```diff
@@ -0,0 +1,32 @@
+module BikeFinder
+  # Find a bike with the given serial number `serial` associated with email
+  # address `owner_email`.
+  #
+  # Matches based on the normalized serial number, and `owner_email` should be
+  # found via:
+  #
+  # - the bike's owner_email attribute
+  # - the email attribute of any of the bike's owner's (user / creator: via bike.ownerships)
+  # - any email associated with any owner (via user.user_emails)
+  #
+  # Return a Bike object, or nil
+  def self.find_matching(serial:, owner_email:)
```

Comment:
Questions about this class:

- Why not use `User.friendly_id_find`?
- What about bikes with absent serials?

Reply:
> Why not use User.friendly_id_find?

I was trying to minimize db calls, but `friendly_id_find` will trigger a query and instantiate a model. That said, I didn't realize we had an `EmailNormalizer`, so I'll add that in here — it'll still be a single query, but it'll match more robustly. Good call-out, thanks.

> What about bikes with absent serials?

Discussed offline: I'll make an issue for handling the `absent` serial corner cases 👍

Reply:
Is there a way to move this sequel logic into `User.friendly_id_find`?

Reply:
Overall performance wouldn't seem to be impacted too much:
```
# Bike Load (47.6ms)  SELECT  "bikes".* FROM "bikes" LEFT JOIN ownerships ON bikes.id = ownerships.bike_id WHERE "bikes"."example" = $1 AND "bikes"."hidden" = $2 AND "bikes"."serial_normalized" = $3 AND (bikes.owner_email = 'elvis8tcb@yahoo.com' OR ownerships.user_id IN (SELECT DISTINCT "users"."id" FROM "users" LEFT JOIN user_emails ON user_emails.user_id = users.id WHERE (users.email = 'elvis8tcb@yahoo.com' OR user_emails.email = 'elvis8tcb@yahoo.com')) OR ownerships.creator_id IN (SELECT DISTINCT "users"."id" FROM "users" LEFT JOIN user_emails ON user_emails.user_id = users.id WHERE (users.email = 'elvis8tcb@yahoo.com' OR user_emails.email = 'elvis8tcb@yahoo.com')))  ORDER BY listing_order desc LIMIT 1  [["example", "f"], ["hidden", "f"], ["serial_normalized", "F6110677"]]
```
vs. 
```
# UserEmail Load (0.2ms)  SELECT  "user_emails".* FROM "user_emails" WHERE (confirmation_token IS NULL) AND "user_emails"."email" = $1 LIMIT 1  [["email", "elvis8tcb@yahoo.com"]]
# User Load (0.5ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT 1  [["id", 145]]
# Bike Load (48.5ms)  SELECT  "bikes".* FROM "bikes" LEFT JOIN ownerships ON bikes.id = ownerships.bike_id WHERE "bikes"."example" = $1 AND "bikes"."hidden" = $2 AND "bikes"."serial_normalized" = $3 AND (bikes.owner_email = 'elvis8tcb@yahoo.com' OR ownerships.user_id IN (145) OR ownerships.creator_id IN (145))  ORDER BY listing_order desc LIMIT 1  [["example", "f"], ["hidden", "f"], ["serial_normalized", "F6110677"]]
```

But:
(1) the result set might be different (`friendly_id_find` only queries in `UserEmail`, not `User`, which has its own `email` column, and restricts results to confirmed emails)
(2) `friendly_id_find` and the methods it depends on would have to be refactored to return an `ActiveRecord::Relation` (rather than an Integer or nil)

Reply:
Shouldn’t both methods search the same thing?

Reply:
> Shouldn’t both methods search/return the same thing?

I can't speak to the intent of `User.friendly_id_find`, but I would think that for the purposes of upserting, we would not want to restrict ourselves to confirmed emails, no? 

(i.e., I'm not sure I see why we would only want to upsert for bikes with an associated user that has a confirmed email—and create duplicates for others—given that we want to upsert for bikes with no associated user yet. Let me know if there's a faulty assumption in there though.)

There's also the `users.email` column, which contains emails not found in `user_emails.emails`.

Reply:
I think the `users.emails` column is guaranteed to be covered by user_emails emails. 

But I think your logic is correct - re: unconfirmed emails.

So you can ignore this

Reply:
I thought there might be application logic enforcing that duplication (in general I'm a little wary of having application logic enforce database constraints), but I'm seeing a significant number of emails on `users` that aren't on `user_emails`:

```rb
User.connection.execute("SELECT COUNT(*) FROM users AS u WHERE u.email NOT IN (SELECT ue.email from user_emails AS ue WHERE ue.user_id = u.id)").to_a.first
# (174.3ms)  SELECT COUNT(*) FROM users AS u WHERE u.email NOT IN (SELECT ue.email from user_emails AS ue WHERE ue.user_id = u.id)
# {
#  "count" => "24471"
# }
```

```ruby
Set.new(User.pluck(:email)).difference(UserEmail.pluck(:email).flatten).count
#   (47.7ms)  SELECT "users"."email" FROM "users"
#   (24.8ms)  SELECT "user_emails"."email" FROM "user_emails"
# 24381
```

```ruby
User.find_by(email: "mary8284@gmail.com").user_emails
```

Possibly an artifact of anonymization?

Reply:
Hmmmm. are those unconfirmed users? Looks like there are 150,125 confirmed vs 124,961 unconfirmed

Reply:
Almost all, from the looks of it -- there's only 12 user emails not in users_emails once I limit the query to confirmed users.